generate/generate.go

```diff
@@ -141,6 +98,93 @@ func Do(root string) error {
 	return nil
 }
 
+// Outdated represents generated code considered outdated, with info
+// detailing where the code is, which stack it belongs, etc.
+type Outdated struct {
+	// StackDir is the dir of the stack.
+	StackDir string
+
+	// Filename is the file name of the outdated code
+	Filename string
+}
+
+// Check will walk all the directories starting from project's root
+// checking if generated code is updated.
+//
+// It will return a list of outdated code descriptions, one for each
+// outdated generated code found. If all code is updated the list will be empty.
+//
+// The provided root must be the project's root directory as an absolute path.
+func Check(root string) ([]Outdated, error) {
+	outdated := []Outdated{}
+	errs := iterateStacks(root, func(
+		stackpath string,
+		metadata terramate.StackMetadata,
+		globals *terramate.Globals,
+	) error {
+		logger := log.With().
+			Str("action", "generate.Check()").
+			Str("path", root).
+			Str("stack", stackpath).
+			Logger()
+
+		logger.Trace().Msg("Checking for outdated backend cfg code on stack.")
+
+		logger.Trace().Msg("Generating backend cfg code for stack.")
+
+		genbackend, err := generateBackendCfgCode(root, stackpath, metadata, globals, stackpath)
+		if err != nil {
+			return err
+		}
+
+		// TODO(katcipis): allow BackendCfgFilename to be configured
+		stackBackendCfgFile := filepath.Join(stackpath, BackendCfgFilename)
+		currentbackend, err := loadGeneratedCode(stackBackendCfgFile)
+		if err != nil {
+			return err
+		}
+
+		if string(genbackend) != string(currentbackend) {
+			logger.Trace().Msg("Detected outdated backend config.")
+			outdated = append(outdated, Outdated{
+				StackDir: metadata.Path,
+				Filename: BackendCfgFilename,
+			})
+		}
+
+		logger.Trace().Msg("Checking for outdated exported locals code on stack.")
+
+		genlocals, err := generateStackLocalsCode(root, stackpath, metadata, globals)
+		if err != nil {
+			return err
+		}
+
+		// TODO(katcipis): allow LocalsFilename to be configured
+		stackLocalsFile := filepath.Join(stackpath, LocalsFilename)
+		currentlocals, err := loadGeneratedCode(stackLocalsFile)
+		if err != nil {
+			return err
+		}
+
+		if string(genlocals) != string(currentlocals) {
+			logger.Trace().Msg("Detected outdated exported locals.")
+			outdated = append(outdated, Outdated{
+				StackDir: metadata.Path,
+				Filename: LocalsFilename,
+			})
+		}
+
+		return nil
+	})
+
+	if len(errs) > 0 {
+		err := fmt.Errorf("checking outdated code: %v : and %d other errors", errs[0], len(errs)-1)
```

Comment:
Inspired on: https://github.com/golang/go/blob/3d3f5d912b776424e50be276bc3a4ae02bf8d143/src/go/scanner/errors.go#L93

But in a more hacky way, maybe we need to model something like this ErrorList from Go (hcl.Diagnose is a worse name IMHO, but it is quite similar, including the Error implementation which is almost identical =P).

Reply:
yeah, at least here it is internal to the function and I convert it to a single err instance, that is why I'm OK with it =P . What would you suggest ?

Reply:
overall for dealing with N operations, where some may fail and some may not, like they are orthogonal to each other, we could just return a slice of Results, and then each result may have an error (failed) or success (err will be nil). It is one of the ideas I had one of these days for generate...but Im not sure and don't wanted to rock the boat with this PR :laughing:

Reply:
it's ok how it's right now. I meant the idea of having `if err.HasError()` or something like that...